## 1. Importing Libraries and Loading the Dataset

In [115]:
## Importing all the necessary libraries
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

In [116]:
## Loading the Heart Dataset
heart_df = pd.read_csv('data/heart.csv')
heart_df

,Age,Sex,ChestPainType,RestingBP,Cholesterol,FastingBS,RestingECG,MaxHR,ExerciseAngina,Oldpeak,ST_Slope,HeartDisease
0,40,M,ATA,140,289,0,Normal,172,N,0.0,Up,0
1,49,F,NAP,160,180,0,Normal,156,N,1.0,Flat,1
2,37,M,ATA,130,283,0,ST,98,N,0.0,Up,0
3,48,F,ASY,138,214,0,Normal,108,Y,1.5,Flat,1
4,54,M,NAP,150,195,0,Normal,122,N,0.0,Up,0
...,...,...,...,...,...,...,...,...,...,...,...,...
913,45,M,TA,110,264,0,Normal,132,N,1.2,Flat,1
914,68,M,ASY,144,193,1,Normal,141,N,3.4,Flat,1
915,57,M,ASY,130,131,0,Normal,115,Y,1.2,Flat,1
916,57,F,ATA,130,236,0,LVH,174,N,0.0,Flat,1


In [117]:
## Displaying 5 Random samples of cataset
heart_df.sample(5)

,Age,Sex,ChestPainType,RestingBP,Cholesterol,FastingBS,RestingECG,MaxHR,ExerciseAngina,Oldpeak,ST_Slope,HeartDisease
243,43,F,ATA,120,266,0,Normal,118,N,0.0,Up,0
503,51,M,ASY,132,227,1,ST,138,N,0.2,Up,0
250,44,M,ASY,135,491,0,Normal,135,N,0.0,Flat,1
712,58,M,ASY,100,234,0,Normal,156,N,0.1,Up,1
419,55,M,ASY,142,228,0,ST,149,Y,2.5,Up,1


In [118]:
## Getting dataset summary
heart_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 918 entries, 0 to 917
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             918 non-null    int64  
 1   Sex             918 non-null    object 
 2   ChestPainType   918 non-null    object 
 3   RestingBP       918 non-null    int64  
 4   Cholesterol     918 non-null    int64  
 5   FastingBS       918 non-null    int64  
 6   RestingECG      918 non-null    object 
 7   MaxHR           918 non-null    int64  
 8   ExerciseAngina  918 non-null    object 
 9   Oldpeak         918 non-null    float64
 10  ST_Slope        918 non-null    object 
 11  HeartDisease    918 non-null    int64  
dtypes: float64(1), int64(6), object(5)
memory usage: 86.2+ KB


In [119]:
## Getting dataset statistics
heart_df.describe().transpose()

,count,mean,std,min,25%,50%,75%,max
Age,918.0,53.510893,9.432617,28.0,47.00,54.0,60.0,77.0
RestingBP,918.0,132.396514,18.514154,0.0,120.00,130.0,140.0,200.0
Cholesterol,918.0,198.799564,109.384145,0.0,173.25,223.0,267.0,603.0
FastingBS,918.0,0.233115,0.423046,0.0,0.00,0.0,0.0,1.0
MaxHR,918.0,136.809368,25.460334,60.0,120.00,138.0,156.0,202.0
Oldpeak,918.0,0.887364,1.066570,-2.6,0.00,0.6,1.5,6.2
HeartDisease,918.0,0.553377,0.497414,0.0,0.00,1.0,1.0,1.0


In [120]:
heart_df.describe(include='all').transpose()

,count,unique,top,freq,mean,std,min,25%,50%,75%,max
Age,918.0,NaN,NaN,NaN,53.510893,9.432617,28.0,47.0,54.0,60.0,77.0
Sex,918,2,M,725,NaN,NaN,NaN,NaN,NaN,NaN,NaN
ChestPainType,918,4,ASY,496,NaN,NaN,NaN,NaN,NaN,NaN,NaN
RestingBP,918.0,NaN,NaN,NaN,132.396514,18.514154,0.0,120.0,130.0,140.0,200.0
Cholesterol,918.0,NaN,NaN,NaN,198.799564,109.384145,0.0,173.25,223.0,267.0,603.0
FastingBS,918.0,NaN,NaN,NaN,0.233115,0.423046,0.0,0.0,0.0,0.0,1.0
RestingECG,918,3,Normal,552,NaN,NaN,NaN,NaN,NaN,NaN,NaN
MaxHR,918.0,NaN,NaN,NaN,136.809368,25.460334,60.0,120.0,138.0,156.0,202.0
ExerciseAngina,918,2,N,547,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Oldpeak,918.0,NaN,NaN,NaN,0.887364,1.06657,-2.6,0.0,0.6,1.5,6.2


## 2. Data Preprocessing

In [121]:
## Checking the NUll Values
heart_df.isna().sum()

Age               0
Sex               0
ChestPainType     0
RestingBP         0
Cholesterol       0
FastingBS         0
RestingECG        0
MaxHR             0
ExerciseAngina    0
Oldpeak           0
ST_Slope          0
HeartDisease      0
dtype: int64

In [122]:
## Checking the duplicates
heart_df.duplicated().sum()

np.int64(0)

In [123]:
## Checking number of unique values in each features
heart_df.nunique()

Age                50
Sex                 2
ChestPainType       4
RestingBP          67
Cholesterol       222
FastingBS           2
RestingECG          3
MaxHR             119
ExerciseAngina      2
Oldpeak            53
ST_Slope            3
HeartDisease        2
dtype: int64

In [124]:
heart_df.columns

Index(['Age', 'Sex', 'ChestPainType', 'RestingBP', 'Cholesterol', 'FastingBS',
       'RestingECG', 'MaxHR', 'ExerciseAngina', 'Oldpeak', 'ST_Slope',
       'HeartDisease'],
      dtype='object')

In [125]:
heart_df

,Age,Sex,ChestPainType,RestingBP,Cholesterol,FastingBS,RestingECG,MaxHR,ExerciseAngina,Oldpeak,ST_Slope,HeartDisease
0,40,M,ATA,140,289,0,Normal,172,N,0.0,Up,0
1,49,F,NAP,160,180,0,Normal,156,N,1.0,Flat,1
2,37,M,ATA,130,283,0,ST,98,N,0.0,Up,0
3,48,F,ASY,138,214,0,Normal,108,Y,1.5,Flat,1
4,54,M,NAP,150,195,0,Normal,122,N,0.0,Up,0
...,...,...,...,...,...,...,...,...,...,...,...,...
913,45,M,TA,110,264,0,Normal,132,N,1.2,Flat,1
914,68,M,ASY,144,193,1,Normal,141,N,3.4,Flat,1
915,57,M,ASY,130,131,0,Normal,115,Y,1.2,Flat,1
916,57,F,ATA,130,236,0,LVH,174,N,0.0,Flat,1


Cholesterol cannot be 0, it states that the reading was not properly noted

#### Change columns type to int

In [126]:
heart_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 918 entries, 0 to 917
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             918 non-null    int64  
 1   Sex             918 non-null    object 
 2   ChestPainType   918 non-null    object 
 3   RestingBP       918 non-null    int64  
 4   Cholesterol     918 non-null    int64  
 5   FastingBS       918 non-null    int64  
 6   RestingECG      918 non-null    object 
 7   MaxHR           918 non-null    int64  
 8   ExerciseAngina  918 non-null    object 
 9   Oldpeak         918 non-null    float64
 10  ST_Slope        918 non-null    object 
 11  HeartDisease    918 non-null    int64  
dtypes: float64(1), int64(6), object(5)
memory usage: 86.2+ KB


In [127]:
## Changing the datatypes
heart_df['RestingBP'] = heart_df['RestingBP'].astype('int64')
heart_df['Cholesterol'] = heart_df['Cholesterol'].astype('int64')
heart_df['FastingBS'] = heart_df['FastingBS'].astype('object')

In [128]:
heart_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 918 entries, 0 to 917
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             918 non-null    int64  
 1   Sex             918 non-null    object 
 2   ChestPainType   918 non-null    object 
 3   RestingBP       918 non-null    int64  
 4   Cholesterol     918 non-null    int64  
 5   FastingBS       918 non-null    object 
 6   RestingECG      918 non-null    object 
 7   MaxHR           918 non-null    int64  
 8   ExerciseAngina  918 non-null    object 
 9   Oldpeak         918 non-null    float64
 10  ST_Slope        918 non-null    object 
 11  HeartDisease    918 non-null    int64  
dtypes: float64(1), int64(5), object(6)
memory usage: 86.2+ KB


#### Imputing the 0 values in Cholesterol column with KNN Imputer

In [129]:
heart_df['Cholesterol'].value_counts()

Cholesterol
0      172
254     11
223     10
220     10
230      9
      ... 
392      1
316      1
153      1
466      1
131      1
Name: count, Length: 222, dtype: int64

In [130]:
## Replacing all the 0 values to the nan value in Cholesterol columns
heart_df['Cholesterol'].replace(0, np.nan, inplace=True)

In [131]:
from sklearn.impute import KNNImputer
knn = KNNImputer(n_neighbors=3)
heart_df['Cholesterol'] = knn.fit_transform(heart_df[['Cholesterol']])

In [132]:
heart_df['Cholesterol'].value_counts()

Cholesterol
244.635389    172
254.000000     11
223.000000     10
220.000000     10
230.000000      9
             ... 
392.000000      1
316.000000      1
153.000000      1
466.000000      1
131.000000      1
Name: count, Length: 222, dtype: int64

In [133]:
heart_df['RestingBP'].value_counts()

RestingBP
120    132
130    118
140    107
110     58
150     55
      ... 
185      1
98       1
92       1
113      1
164      1
Name: count, Length: 67, dtype: int64

In [134]:
## Replacing all the 0 values to the nan value in RestingBP columns
heart_df['RestingBP'].replace(0, np.nan, inplace=True)

In [135]:
from sklearn.impute import KNNImputer
knn = KNNImputer(n_neighbors=3)
heart_df['RestingBP'] = knn.fit_transform(heart_df[['RestingBP']])

In [136]:
heart_df['RestingBP'].unique()

array([140.        , 160.        , 130.        , 138.        ,
       150.        , 120.        , 110.        , 136.        ,
       115.        , 100.        , 124.        , 113.        ,
       125.        , 145.        , 112.        , 132.        ,
       118.        , 170.        , 142.        , 190.        ,
       135.        , 180.        , 108.        , 155.        ,
       128.        , 106.        ,  92.        , 200.        ,
       122.        ,  98.        , 105.        , 133.        ,
        95.        ,  80.        , 137.        , 185.        ,
       165.        , 126.        , 152.        , 116.        ,
       132.54089422, 144.        , 154.        , 134.        ,
       104.        , 139.        , 131.        , 141.        ,
       178.        , 146.        , 158.        , 123.        ,
       102.        ,  96.        , 143.        , 172.        ,
       156.        , 114.        , 127.        , 101.        ,
       174.        ,  94.        , 148.        , 117.  

In [137]:
heart_df['RestingBP'].isna().sum()

np.int64(0)

## 3. Data Visualization

In [138]:
import plotly.express as px

In [139]:
heart_df.sample()

,Age,Sex,ChestPainType,RestingBP,Cholesterol,FastingBS,RestingECG,MaxHR,ExerciseAngina,Oldpeak,ST_Slope,HeartDisease
871,61,M,NAP,150.0,243.0,1,Normal,137,Y,1.0,Flat,0


In [140]:
heart_df.select_dtypes(include=['int', 'float'])

,Age,RestingBP,Cholesterol,MaxHR,Oldpeak,HeartDisease
0,40,140.0,289.0,172,0.0,0
1,49,160.0,180.0,156,1.0,1
2,37,130.0,283.0,98,0.0,0
3,48,138.0,214.0,108,1.5,1
4,54,150.0,195.0,122,0.0,0
...,...,...,...,...,...,...
913,45,110.0,264.0,132,1.2,1
914,68,144.0,193.0,141,3.4,1
915,57,130.0,131.0,115,1.2,1
916,57,130.0,236.0,174,0.0,1


In [141]:
numerical_columns = heart_df.select_dtypes(include=['int', 'float'])

numerical_columns.corr()['HeartDisease'][:-1].sort_values()

MaxHR         -0.400421
Cholesterol    0.094071
RestingBP      0.117938
Age            0.282039
Oldpeak        0.403951
Name: HeartDisease, dtype: float64

In [142]:
corr_values = numerical_columns.corr()['HeartDisease'][:-1].sort_values()

fig = px.bar(
    x=corr_values.index,
    y=corr_values.values,
    title='Correlation with HeartDisease',
    labels={'x': 'Features', 'y': 'Correlation'},
    color=corr_values.values,
    color_continuous_scale='RdBu'
)

fig.show()

#### Age and HeartDisease Distribution

In [143]:
px.sunburst(heart_df, path=['HeartDisease', 'Age'])

In [144]:
px.histogram(heart_df, x='Age', color='HeartDisease')

#### Percentage of HeartDisease data distribution

In [145]:
## Percentage of HeartDisease data distribution
px.pie(heart_df, names='HeartDisease', title='Percentage os HeartDisease classes distribution')

#### Sex vs HeartDisease

In [146]:
px.histogram(heart_df, x='Sex', color='HeartDisease')

* Female Patient Having low chances of having the Heart Disease

#### ChestPainType vs HeartDisease

In [147]:
px.histogram(heart_df, x='ChestPainType', color='HeartDisease')

* ASY ChesPainType Patiest having high chances of being heart disease
* ATA ChesPainType Patiest having low chances of being heart disease

#### RestingBP vs HeartDisease

In [148]:
heart_df['RestingBP'].unique()

array([140.        , 160.        , 130.        , 138.        ,
       150.        , 120.        , 110.        , 136.        ,
       115.        , 100.        , 124.        , 113.        ,
       125.        , 145.        , 112.        , 132.        ,
       118.        , 170.        , 142.        , 190.        ,
       135.        , 180.        , 108.        , 155.        ,
       128.        , 106.        ,  92.        , 200.        ,
       122.        ,  98.        , 105.        , 133.        ,
        95.        ,  80.        , 137.        , 185.        ,
       165.        , 126.        , 152.        , 116.        ,
       132.54089422, 144.        , 154.        , 134.        ,
       104.        , 139.        , 131.        , 141.        ,
       178.        , 146.        , 158.        , 123.        ,
       102.        ,  96.        , 143.        , 172.        ,
       156.        , 114.        , 127.        , 101.        ,
       174.        ,  94.        , 148.        , 117.  

In [149]:
px.sunburst(heart_df, path=['HeartDisease', 'RestingBP'])

#### FastingBS vs Heart Disease

In [150]:
px.histogram(heart_df, x='FastingBS', color='HeartDisease')

* The patient having Blood Sugar count greater than 120 having High chances of being Heart Disease

#### MaxHR vs HeartDisease

In [151]:
px.sunburst(heart_df, path=['HeartDisease', 'MaxHR'])

In [152]:
px.violin(heart_df, x='HeartDisease', y='MaxHR', color='HeartDisease')

* Higher Heart Rate Person having the less chances of Heart Disease

#### OldPeak vs Heart Disease

In [153]:
px.violin(heart_df, x='HeartDisease', y='Oldpeak', color='HeartDisease')

* Patient with Oldpeak near to zero having low chances of having the Heart Disease

#### ST_Slop vs Heart Disease

In [154]:
px.histogram(heart_df, x='ST_Slope', color='HeartDisease')

* Up slope patient have High chances of geeting Heart Disease


* Flat slope patient have High chances of geeting Heart Disease

* Down slope patient have High chances of not geeting Heart Disease


#### ExerciseAngina vs Heart Disease

In [155]:
px.histogram(heart_df, x='ExerciseAngina', color='HeartDisease')

* Patient with ExerciseAngina having High chances of having Heart Disease

#### RestingECG vs Heart Disease

In [156]:
px.histogram(heart_df, x='RestingECG', color='HeartDisease')

# 4. Model Training

## Train Test Split

In [157]:
heart_df.select_dtypes(include='object').nunique()

Sex               2
ChestPainType     4
FastingBS         2
RestingECG        3
ExerciseAngina    2
ST_Slope          3
dtype: int64

In [158]:
X = heart_df.drop(columns=['HeartDisease'], axis=1)
y = heart_df['HeartDisease']

In [159]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=heart_df['HeartDisease'])

In [160]:
from sklearn.impute import KNNImputer,SimpleImputer
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

In [161]:
num_col = X_train.select_dtypes(include='number').columns.tolist()
cat_col = X_train.select_dtypes(include='object').columns.tolist()
target_col = heart_df[['HeartDisease']]

In [162]:
num_pipeline = Pipeline(steps=[
    ('imputer', KNNImputer(n_neighbors=3)),
    ('scaler', StandardScaler())
])


cat_pipeline = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('encode', OneHotEncoder())
])

In [163]:
preprocessor = ColumnTransformer(transformers=[
    ('num_pipe', num_pipeline, num_col),
    ('cat_pipe', cat_pipeline, cat_col)
])

In [164]:
X_train_transformed = preprocessor.fit_transform(X_train)
X_test_transformed = preprocessor.transform(X_test)

## Logistic Regression

In [165]:
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import accuracy_score, f1_score
from sklearn.linear_model import LogisticRegression

lr_param_grid = {
    'solver': ['lbfgs', 'liblinear', 'newton-cg', 'newton-cholesky', 'sag', 'saga']
}

lr_grid = GridSearchCV(LogisticRegression(), lr_param_grid, cv=5, scoring='accuracy')
lr_grid.fit(X_train_transformed, y_train)

lr_pred = lr_grid.best_estimator_.predict(X_test_transformed)
print(f'LogisticRegression Best Solver : {lr_grid.best_params_}')
print(f'LogisticRegression Accuracy    : {accuracy_score(y_test, lr_pred)}')
print()

LogisticRegression Best Solver : {'solver': 'lbfgs'}
LogisticRegression Accuracy    : 0.8858695652173914



In [166]:
import pickle

with open('artifacts/LogisticRegression.pkl', 'wb') as model:
    pickle.dump(lr_grid, model) 

## Support Vector Machine

In [167]:
from sklearn.svm import SVC

svm_param_grid = {
    'kernel': ['linear', 'poly', 'rbf', 'sigmoid']
}

svm_grid = GridSearchCV(SVC(), svm_param_grid, cv=5, scoring='f1_weighted')
svm_grid.fit(X_train_transformed, y_train)

svm_pred = svm_grid.best_estimator_.predict(X_test_transformed)
print(f'SVM Best Kernel : {svm_grid.best_params_}')
print(f'SVM F1 Score    : {f1_score(y_test, svm_pred, average="weighted")}')
print()

SVM Best Kernel : {'kernel': 'rbf'}
SVM F1 Score    : 0.8689519616322038



In [168]:
with open('artifacts/SVM.pkl', 'wb') as model:
    pickle.dump(svm_grid, model)

## Decision Tree Classifier

In [169]:
from sklearn.tree import DecisionTreeClassifier

dt_param_grid = {
    'max_depth': [3, 4, 5, 6, 7, 8],
    'min_samples_split': [2, 3, 4],
    'min_samples_leaf': [1, 2, 3, 4],
    'random_state': [0, 42]
}

dt_grid = GridSearchCV(DecisionTreeClassifier(class_weight='balanced'), dt_param_grid, cv=5, scoring='accuracy')
dt_grid.fit(X_train_transformed, y_train)

dt_pred = dt_grid.best_estimator_.predict(X_test_transformed)
print(f'DecisionTree Best Params : {dt_grid.best_params_}')
print(f'DecisionTree Accuracy    : {accuracy_score(y_test, dt_pred)}')
print()

DecisionTree Best Params : {'max_depth': 3, 'min_samples_leaf': 4, 'min_samples_split': 2, 'random_state': 0}
DecisionTree Accuracy    : 0.8206521739130435



In [170]:
with open('artifacts/DecisionTree.pkl', 'wb') as model:
    pickle.dump(dt_grid, model)

## Random Forest Classifier

In [171]:
from sklearn.ensemble import RandomForestClassifier

rf_param_grid = {
    'n_estimators': [50, 100, 150, 500],
    'max_features': ['sqrt', 'log2', None],
    'max_depth': [3, 6, 9, 19],
    'max_leaf_nodes': [3, 6, 9]
}

rf_grid = GridSearchCV(RandomForestClassifier(), rf_param_grid, cv=5, scoring='accuracy')
rf_grid.fit(X_train_transformed, y_train)

rf_pred = rf_grid.best_estimator_.predict(X_test_transformed)
print(f'RandomForest Best Params : {rf_grid.best_params_}')
print(f'RandomForest Accuracy    : {accuracy_score(y_test, rf_pred)}')

RandomForest Best Params : {'max_depth': 19, 'max_features': 'log2', 'max_leaf_nodes': 9, 'n_estimators': 50}
RandomForest Accuracy    : 0.8804347826086957


In [172]:
with open('artifacts/RandomForest.pkl', 'wb') as model:
    pickle.dump(rf_grid, model)